src/core/common/iterator_utils.hpp

```diff
@@ -0,0 +1,152 @@
+/*
+ *  Copyright (c) 2020, The OpenThread Authors.
+ *  All rights reserved.
+ *
+ *  Redistribution and use in source and binary forms, with or without
+ *  modification, are permitted provided that the following conditions are met:
+ *  1. Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *  2. Redistributions in binary form must reproduce the above copyright
+ *     notice, this list of conditions and the following disclaimer in the
+ *     documentation and/or other materials provided with the distribution.
+ *  3. Neither the name of the copyright holder nor the
+ *     names of its contributors may be used to endorse or promote products
+ *     derived from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ *  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ *  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ *  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ *  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ *  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ *  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * @file
+ *   This file includes definitions for a generic item-pointer iterator class.
+ */
+
+#ifndef ITERATOR_UTILS_HPP_
+#define ITERATOR_UTILS_HPP_
+
+namespace ot {
+
+/**
+ * @addtogroup core-iterator-utils
+ *
+ * @brief
+ *   This module includes definitions for OpenThread generic item-pointer iterator class.
+ *
+ * @{
+ *
+ */
+
+/**
+ * This template class is used as a base class for those item-pointer iterators.
+ *
+ * These iterators have common methods and operators like `Advance()` and `++` and hold a pointer to the
+ * object.
+ *
+ * Users of this class should follow CRTP-style inheritance, i.e., `IteratorType` class itself should publicly inherit
+ * from `ItemPtrIterator<ItemType, IteratorType>`. `ItemType` is the type of the object that `IteratorType` points to.
+ *
+ * The template type `IteratorType` should contain a `mItem` member variable  which is a pointer of `ItemType`,
+ * and have a method `Advance()`.
+ *
+ */
+template <class ItemType, class IteratorType> class ItemPtrIterator
```

Comment:
One idea we can consider that may make this a bit simpler to just have one template parameter `IteratorType`.

Then we expect/require the `ItemType`  to be define by `IteratorType` itself (i.e., be a nested `typedef` in it), same way that we require it to have `Advance()`method.

In the implementation here, we can simply refer to it a `IteratorType::ItemType`. 

I also like the idea of having `mItem` be defined by this class itself.

Thoughts?

Reply:
Sounds good to me, thanks!